src/3d/qgsrulebasedchunkloader_p.cpp

```diff
@@ -203,4 +204,55 @@ void QgsRuleBasedChunkedEntity::onTerrainElevationOffsetChanged( float newOffset
   mTransform->setTranslation( QVector3D( 0.0f, newOffset, 0.0f ) );
 }
 
+QVector<RayHit> QgsRuleBasedChunkedEntity::rayIntersection( const QgsRay3D &ray, const RayCastContext &context ) const
+{
+  Q_UNUSED( context )
+  QgsDebugMsgLevel( QStringLiteral( "Ray cast on vector layer" ), 2 );
+  int nodeUsed = 0;
+  int nodesAll = 0;
+  int hits = 0;
+  QVector<RayHit> result;
+
+  float minDist = -1;
+  QVector3D intersectionPoint;
+  QgsFeatureId fid;
+
+  const QList<QgsChunkNode *> activeNodes = this->activeNodes();
+  for ( QgsChunkNode *node : activeNodes )
+  {
+    nodesAll++;
+    if ( node->entity() &&
+         ( minDist < 0 || node->bbox().distanceFromPoint( ray.origin() ) < minDist ) &&
+         ray.intersects( Qgs3DUtils::aabbToBox( node->bbox() ) ) )
+    {
+      nodeUsed++;
+      Qt3DRender::QGeometryRenderer *rend = node->entity()->findChild<Qt3DRender::QGeometryRenderer *>();
+      Qt3DRender::QGeometry *geom = rend->geometry();
+      QgsTessellatedPolygonGeometry *polygonGeom = qobject_cast<QgsTessellatedPolygonGeometry *>( geom );
+      if ( !polygonGeom )
+        return result; // other QGeometry types are not supported for now
+      Qt3DCore::QTransform *tr = node->entity()->findChild<Qt3DCore::QTransform *>();
+      const QMatrix4x4 matrix = tr ? tr->matrix() : QMatrix4x4();
+
+      QVector3D nodeIntPoint;
+      if ( polygonGeom->rayIntersection( ray, matrix, nodeIntPoint, fid ) )
+      {
+        hits++;
+        float dist = ( ray.origin() - nodeIntPoint ).length();
+        if ( minDist < 0 || dist < minDist )
```

Comment:
micro optimization: it is often useful to initialize `minDist` to some huge value, so that you do not need to worry about the special case of the first hit...